keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,549 @@
+        backoffLimit: 0 
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: driver
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 100
+  - name: JobB
+    replicas: 1
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+  - name: JobC
+    replicas: 1
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+In the JobSet API, ReplicatedJobs are the individual Job elements that are created.  These are Replicated so you can have the same Job replicated.  This adds some complexity in the concept of Ready as we must consider the overall ReplicatedJob Ready rather than using the individual Jobs.
+
+For example, a Job with a `Parallelism`: 4 and `Ready` of 3. Means that 1 pod of a Job is not ready.  
+The Job API has no concept of an overall readiness condition based on a percentage of pods that are ready.  
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API Proposal
+
+```golang
+// Status of ReplicatedJob for StartupPolicy
+type Status string
+
+const (
+ // Ready means that the job is ready
+ Ready Status = "Ready"
+
+ // Succeeded means that the job is successful
+ Succeeded Status = "Succeeded"
+)
+
+type StartupPolicy struct {
+ // TargetReplicatedJobStatuses is a list of rules for applying StartupPolicy.
+ // +optional
+ // +listType=map
+ // +listMapKey=name
+ TargetReplicatedJobStatuses []TargetReplicatedJobStartup `json:"targetReplicatedJobStatuses,omitempty"`
+}
+
+type TargetReplicatedJobStartup struct {
+ // Name of ReplicatedJob to apply startup policy
+ Name string `json:"name"`
+
+ // The status you want the ReplicatedJob to be
+ // +kubebuilder:validation:Enum=Ready;Succeeded
+ Status Status `json:"status"`
+
+ // Number of Statuses is how many ReplicatedJob statuses to consider "started"
+ // Empty will mean all status must match the replicas.
```

Comment:
The name is confusing, it is not clear to me what "number of statuses" means? statuses isn't something we usually count

Reply:
Yea I originally thought we could consider the Job to report the ready status (and use a percentage of Jobs) to decide when a job was ready.  That wasn't accepted in upstream so I decided to try out this idea.

Reply:
Updated this.